### PR TITLE
Fixed #27142, #27110 -- Made makemigrations consistency checks respect database routers.

### PR DESCRIPTION
**Pull request by timgraham**
Partially reverted refs #27054 except for one of the tests as this solution supersedes that one.

https://code.djangoproject.com/ticket/27142
https://code.djangoproject.com/ticket/27110
https://code.djangoproject.com/ticket/27054

### CODE REVIEW

**Comment by SydneyUniLibrary-Jim** [comment]
For my case, reported in #27054, this patch works for me.

I did have to change my router code. The models for my 3rd party database are isolated into their own app. In my allow_migrate I was returning False if the app_label was that of the app for the 3rd-party database, and None otherwise.

This was working for me in Django 1.9, but I don't think I was actually doing the correct thing.

I am now returning False if the app_label is that of the the app for the 3rd-party database app, or if the db is the 3rd-party database; None otherwise.

File: django/core/management/commands/makemigrations.py [hunk 1]
```diff
@@ -94,9 +95,18 @@ def handle(self, *app_labels, **options):
...
-            if connection.settings_dict['ENGINE'] != 'django.db.backends.dummy':
+            is_default = (db == DEFAULT_DB_ALIAS)
+            # Non-default databases are only checked if database routers used.
+            is_non_default_checkable = (db != DEFAULT_DB_ALIAS and settings.DATABASE_ROUTERS)
+            is_checkable = (
+                (is_default or is_non_default_checkable) and
+                # At least one app must be migrated to the database.
+                any(router.allow_migrate(connection.alias, label) for label in consistency_check_labels)
+            )
+            if connection.settings_dict['ENGINE'] != 'django.db.backends.dummy' and is_checkable:
```
Comment (shaib) [inline django/core/management/commands/makemigrations.py]: This logic seems a little convoluted. Consider:
```python
conns = connetions.values() if settings.DATABASE_ROUTERS else [connections[DEFAULT_DB_ALIAS]]
for conn in conns:
    if (connection.settings_dict['ENGINE'] != 'django.db.backends.dummy' and
        any(router.allow_migrate(connection.alias, label) for label in labels)):
```

File: tests/migrations/test_commands.py [hunk 2]
```diff
@@ -597,26 +598,48 @@ def test_makemigrations_empty_connections(self):
...
+                self.assertEqual(ensure_schema.call_count, 2)  # 'default' again
+
+                # With a router that doesn't prohibit migrating 'other',
+                # consistency is checked.
+                with self.settings(DATABASE_ROUTERS=['migrations.routers.EmptyRouter']):
+                    with self.assertRaisesMessage(MigrationSchemaMissing, 'Patched'):
+                        call_command('makemigrations', 'migrations', verbosity=0)
+                self.assertEqual(ensure_schema.call_count, 4)  # 'default' and 'other'
+
+                # With a router that doesn't prohibit migrating on any database,
```
Comment (shaib) [inline tests/migrations/test_commands.py]: ... that doesn't *allow* migrating ...

**Comment by shaib** [comment]
Other than the minor comments above, I think we should explain the behavior a little in docs. This doesn't necessarily need to delay 1.10.1, though.

File: django/core/management/commands/makemigrations.py (hunk 1 above)
Comment (shaib) [inline django/core/management/commands/makemigrations.py]: actually, one needs to be a little more careful:
```python
conns = (connetions.values() if settings.DATABASE_ROUTERS else
               [connections[DEFAULT_DB_ALIAS]] if DEFAULT_DB_ALIAS in connections else
               [])
```

Comment (timgraham) [inline django/core/management/commands/makemigrations.py]: `ConnectionHandler` always sets up a default database, so I don't think the `if/else` is required.
https://github.com/django/django/blob/7ed8b98f42cd3a9fa530bbb764d3d79230e0a7a7/django/db/utils.py#L157-L162

Comment (shaib) [inline django/core/management/commands/makemigrations.py]: Right (actually there's a bug in that code, the `ImproperlyConfigured` can never be raised).

**Comment by vanflymen** [comment]
I'm using Django 1.10.4 and it looks like `allow_migrate()` is being ignore, and `django_migrations` is trying to be created:

```
django.db.migrations.exceptions.MigrationSchemaMissing: Unable to create the django_migrations table (Column "django_migrations.id" has unsupported type "serial".)
```

**settings.py**
```python
DATABASES = {
    'default': env.db(),
    'redshift': env.db('REDSHIFT_URI')
}
DATABASE_ROUTERS = [
    'my_project.common.routers.CustomRouter',
]
```


**my_project/common/routers.py**
```python
class CustomRouter(object):
    def db_for_read(self, model, **hints):
        return 'default'

    def db_for_write(self, model, **hints):
        return 'default'

    def allow_relation(self, obj1, obj2, **hints):
        db_list = ('default', )
        if obj1._state.db in db_list and obj2._state.db in db_list:
            return True
        return None

    def allow_migrate(self, db, app_label, model_name=None, **hints):
        return db == 'default'
```

**Comment by timgraham** [comment]
The ticket tracker for reporting bugs is https://code.djangoproject.com/.

**Comment by shaib** [comment]
to get to a problem, it looks like you'd have to run `manage.py migrate --database redshift`. Why would you do that?

**Comment by vanflymen** [comment]
@shaib No, I didn't... just `manage.py makemigrations`. 
But, I just discovered (after upgrading from 1.10 -> 1.10.4) that if I don't use a custom router, `makemigrations` doesn't check other defined databases.

**Comment by vanflymen** [comment]
@shaib: Ah, I see what you meant. Redshift is basically importing the RDS MySQL database... including the `django_*` tables.

**Comment by MiddleFork** [comment]
#27110 appears to break #27054 when a read-only database is specified in settings.DATABASES but lacks a corresponding reference in settings.DATABASE_ROUTERS as in dvf's example above.

Makemigrations continues to rely on django.db.utils.ConnectionHandler to supply a set of connection objects populated from settings.DATABASES; those connections are passed along to  loader.check_consistent_history which will once again fail via applied_migrations and ensure_schema.

I added a comment with traceback at #27054

**Pull request merged**